File: Solutions/0725/0725.py
# -*- coding:utf-8 -*-
# Author: washing
# DateTime: 2021/9/22 16:39
# File: 0725.py
# Desc: 

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: ListNode, k: int) -> List[ListNode]:
        l = []
        while head:
            l.append(head.val)
            head = head.next
        n = len(l)
        num = n // k
        ys = n % k
        ret = [None for i in range(k)]
        l_idx = 0
        idx = 0
        while l_idx < k:
            cur = ret[l_idx]
            plus = 0
            if l_idx < ys:
                plus = 1
            for i in range(num+plus):
                if not cur:
                    cur = ret[l_idx] = ListNode(l[idx])
                else:
                    cur.next = ListNode(l[idx])
                    cur = cur.next
                idx += 1
            l_idx += 1
        return ret

File: Solutions/0725/0725.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        int n=0;
        ListNode *cur = head;
        while (cur){
            n++;
            cur = cur -> next;
        }
        int num = n/k, ys = n%k;
        vector<ListNode*> ret(k, nullptr);
        cur = head;
        for(int i=0; i<k && cur; i++){
            ret[i] = cur;
            int plus = num + (i < ys ? 1 : 0);
            for (int j=1; j<plus; j++) cur = cur -> next;
            ListNode *next = cur -> next;
            cur -> next = nullptr;
            cur = next;
        }
        return ret;
    }
};
